rasa/shared/core/domain.py

```diff
@@ -598,9 +600,43 @@ def _collect_overridden_default_intents(
         }
         return sorted(intent_names & set(rasa.shared.core.constants.DEFAULT_INTENTS))
 
-    @staticmethod
+    @classmethod
+    def _update_slots_with_global_not_intent_parameter(
+        cls, form_data: Union[Dict[Text, Any], List[Text]],
+    ) -> Union[Dict[Text, Any], List[Text]]:
+        if "global_not_intent" in form_data:
+            global_params = form_data.get("global_not_intent")
+            # accessing the form slots by using keyword `required_slots`
+            if "required_slots" in form_data:
+                form_data = form_data.get("required_slots")
+            # checking if the global parameter is a list or not and always
+            # making it a list.
+            if not isinstance(global_params, list):
+                global_params = [global_params]
+            # looping through the form's global params.
+            for global_val in global_params:
+                # Updating the `not_intent` parameter of each slot with
+                # `global_not_intent` value.
+                for slot_key, slot_val in form_data.items():
+                    for slot in slot_val:
+                        key = "not_intent"
+                        # check that `not_intent` param is present in the slot
+                        if key in slot.keys():
+                            # check that the value of `not_intent` is a list
+                            if isinstance(slot[key], list):
+                                # check that `global_not_intent` is in the list
+                                if global_val not in slot[key]:
+                                    slot[key].append(global_val)
```

Comment:
Now I know what the downside is of handling this in the domain instead of on the fly in the `FormAction` 😬 When we dump the domain then this will currently alter the domain, e.g.:

```yaml
forms:
  my_form:
    global_not_intent: [greet]
    required_slots:
      name:
      - type: from_text
```

would become
```yaml
forms:
  my_form:
    global_not_intent: [greet]
    required_slots:
      name:
      - type: from_text
        not_intent: [greet]
```

For the Rasa Open Source world this isn't a big deal, but for Integrated Version Control in Rasa X this will lead to big diffs and users might be confused by that. In my opinion we either need to revert the domain as done [here](https://github.com/RasaHQ/rasa/blob/b9705f455d05cc619c222d16970f2b412fae3af1/rasa/shared/core/domain.py#L1326) or we add the global ignored intents in the `FormAction` which seems a tad simpler to me.

Reply:
@wochinge I agree with the `FormAction` solution! 👌🏾 So, I will revert the changes done in the `domain` and I will start implementing it in the `FormAction`.

Reply:
Sorry, for not having thought of this earlier. We need to keep the `required_slots` stuff in the domain. The "only" thing which would need moving to the `FormAction` is the handling of the `global_not_intent` itself (so you'd remove the part from the domain which inserts the `global_not_intent` into each slot mapping)

Reply:
Slick!

Reply:
I pushed some of the changes just to double check with you that I am on the right track. If you could check the changes in the `domain` and `forms` and give me a feedback if that is what you meant, that would be great. 🙏🏽 However in the current implementation I see some issues, for example :
- In the function `get_mappings_for_slot ` I go through all the forms but I should only update (with global_param) the form where the specific slot is (since `get_mappings_for_slot` addresses only to one slot (`slot_to_fill`)). But how do I know in which form I am ? 🧐
- The `get_mappings_for_slot ` looks like it updates only the `from_entity` type in the slot? But I should update the `not_intent` of all the types right (with `global_not_intent`)?

Reply:
Are you sure that you've pushed all changes? I Can't see any within `forms`?